targets/TARGET_Cypress/TARGET_PSOC6/TARGET_WHD/src/bus_protocols/whd_bus_sdio_protocol.c

```diff
@@ -0,0 +1,1427 @@
+#define F2_AVAIL_TIMEOUT_MS   (500)
+#define F2_READY_TIMEOUT_MS   (1000)
+#define ALP_AVAIL_TIMEOUT_MS  (100)
+#define HT_AVAIL_TIMEOUT_MS   (500)
+#define ABORT_TIMEOUT_MS      (100)
+/* Taken from FALCON_5_90_195_26 dhd/sys/dhd_sdio.c. */
+#define SDIO_F2_WATERMARK     (8)
+
+#define INITIAL_READ   4
+
+#define WHD_THREAD_POLL_TIMEOUT      (CY_RTOS_NEVER_TIMEOUT)
+
+#define WHD_THREAD_POKE_TIMEOUT      (100)
+
+#define HOSTINTMASK                 (I_HMB_SW_MASK)
+
+
+/******************************************************
+*             Structures
+******************************************************/
+struct whd_bus_priv
+{
+    whd_sdio_config_t sdio_config;
+    whd_bus_stats_t whd_bus_stats;
+    void *sdio_obj;
+    whd_sdio_funcs_t *sdio_ops;
+
+};
+
+/******************************************************
+*             Variables
+******************************************************/
+
+/******************************************************
+*             Static Function Declarations
+******************************************************/
+
+static whd_result_t whd_bus_sdio_transfer(whd_driver_t whd_driver, whd_bus_transfer_direction_t direction,
+                                          whd_bus_function_t function, uint32_t address, uint16_t data_size,
+                                          uint8_t *data, sdio_response_needed_t response_expected);
+static whd_result_t whd_bus_sdio_cmd52(whd_driver_t whd_driver, whd_bus_transfer_direction_t direction,
+                                       whd_bus_function_t function, uint32_t address, uint8_t value,
+                                       sdio_response_needed_t response_expected, uint8_t *response);
+static whd_result_t whd_bus_sdio_cmd53(whd_driver_t whd_driver, whd_bus_transfer_direction_t direction,
+                                       whd_bus_function_t function, sdio_transfer_mode_t mode, uint32_t address,
+                                       uint16_t data_size, uint8_t *data,
+                                       sdio_response_needed_t response_expected,
+                                       uint32_t *response);
+static whd_result_t whd_bus_sdio_abort_read(whd_driver_t whd_driver, whd_bool_t retry);
+static whd_result_t whd_bus_sdio_download_firmware(whd_driver_t whd_driver);
+
+static whd_result_t whd_bus_sdio_set_oob_interrupt(whd_driver_t whd_driver, uint8_t gpio_pin_number);
+
+static void whd_bus_sdio_irq_handler(void *handler_arg, cyhal_sdio_irq_event_t event);
+static whd_result_t whd_bus_sdio_irq_register(whd_driver_t whd_driver);
+static whd_result_t whd_bus_sdio_irq_enable(whd_driver_t whd_driver, whd_bool_t enable);
+static whd_result_t whd_bus_sdio_init_oob_intr(whd_driver_t whd_driver);
+
+/******************************************************
+*             SDIO Logging
+* Enable this section for logging of SDIO transfers
+* by changing "if 0" to "if 1"
+******************************************************/
+#if 0
```

Comment:
Maybe this should be put behind a configuration flag

Reply:
+1 but looking at the file path, this might is in their driver (3rd party)

Reply:
This code is used internally by WHD driver developers to debug SDIO transfers. There is no value in exposing this as configuration flag for customers.